ipapython/ccache_storage.py

```diff
@@ -0,0 +1,234 @@
+
+    def krb5_errcheck(result, func, arguments):
+        """Error checker for krb5_error return value"""
+        if result != 0:
+            raise KRB5Error(result, func.__name__, arguments)
+
+    krb5_principal = ctypes.POINTER(krb5_principal_data)
+    krb5_context = ctypes.POINTER(_krb5_context)
+    krb5_ccache = ctypes.POINTER(_krb5_ccache)
+    krb5_data_p = ctypes.POINTER(_krb5_data)
+    krb5_error = ctypes.c_int32
+
+    krb5_init_context = LIBKRB5.krb5_init_context
+    krb5_init_context.argtypes = (ctypes.POINTER(krb5_context), )
+    krb5_init_context.restype = krb5_error
+    krb5_init_context.errcheck = krb5_errcheck
+
+    krb5_free_context = LIBKRB5.krb5_free_context
+    krb5_free_context.argtypes = (krb5_context, )
+    krb5_free_context.retval = None
+
+    krb5_free_principal = LIBKRB5.krb5_free_principal
+    krb5_free_principal.argtypes = (krb5_context, krb5_principal)
+    krb5_free_principal.retval = None
+
+    krb5_free_data_contents = LIBKRB5.krb5_free_data_contents
+    krb5_free_data_contents.argtypes = (krb5_context, krb5_data_p)
+    krb5_free_data_contents.retval = None
+
+    krb5_cc_default = LIBKRB5.krb5_cc_default
+    krb5_cc_default.argtypes = (krb5_context, ctypes.POINTER(krb5_ccache), )
+    krb5_cc_default.restype = krb5_error
+    krb5_cc_default.errcheck = krb5_errcheck
+
+    krb5_cc_close = LIBKRB5.krb5_cc_close
+    krb5_cc_close.argtypes = (krb5_context, krb5_ccache, )
+    krb5_cc_close.retval = krb5_error
+    krb5_cc_close.errcheck = krb5_errcheck
+
+    krb5_parse_name = LIBKRB5.krb5_parse_name
+    krb5_parse_name.argtypes = (krb5_context, ctypes.c_char_p,
+                                ctypes.POINTER(krb5_principal), )
+    krb5_parse_name.retval = krb5_error
+    krb5_parse_name.errcheck = krb5_errcheck
+
+    krb5_cc_set_config = LIBKRB5.krb5_cc_set_config
+    krb5_cc_set_config.argtypes = (krb5_context, krb5_ccache, krb5_principal,
+                                   ctypes.c_char_p, krb5_data_p, )
+    krb5_cc_set_config.retval = krb5_error
+    krb5_cc_set_config.errcheck = krb5_errcheck
+
+    krb5_cc_get_config = LIBKRB5.krb5_cc_get_config
+    krb5_cc_get_config.argtypes = (krb5_context, krb5_ccache, krb5_principal,
+                                   ctypes.c_char_p, krb5_data_p, )
+    krb5_cc_get_config.retval = krb5_error
+    krb5_cc_get_config.errcheck = krb5_errcheck
+
+class session_store:
+    def __init__(self, name='X-IPA-Session-Cookie'):
+        self.__context = None
+        if isinstance(LIBKRB5, Exception):  # pragma: no cover
+            raise LIBKRB5
+        context = krb5_context()
+        krb5_init_context(ctypes.byref(context))
+        self.__context = context
+
+        self._hidden_cred_name = name
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, type, value, traceback):
+        if self.__context:
+            krb5_free_context(self.__context)
+            self.__context = None
+
+    def __del__(self):
```

Comment:
Don't define `__del__` method in python unless you have bulletproof evidence that it will work in your case.
https://docs.python.org/2/reference/datamodel.html#object.__del__

In python, if you want safely do cleanup, it must be done by contextmanager (what you have already implemented), don't use __del__ method